### PR TITLE
T5 rpe fix 01

### PR DESCRIPTION
**Pull request by khcs**
# What does this PR do ?

Fix the RPE implementation converging slower with loss consistently higher than NeMo APE and Megatron-LM RPE implementation.

**Collection**: NLP

# Changelog 
- Changed how RPE parameters are initialized and passed on.

# Usage
* Train with `model.position_embedding_type='relative'` config.

```python
python /home/workspace/NeMo/examples/nlp/language_modeling/megatron_t5_pretraining.py model.position_embedding_type='relative'
```

# Before your PR is "Ready for review"
**Pre checks**:
- [x] Make sure you read and followed [Contributor guidelines](https://github.com/NVIDIA/NeMo/blob/main/CONTRIBUTING.md)
- [x] Did you write any new necessary tests?
- [x] Did you add or update any necessary documentation?
- [ ] Does the PR affect components that are optional to install? (Ex: Numba, Pynini, Apex etc)
  - [ ] Reviewer: Does the PR have correct import guards for all optional libraries?
  
**PR Type**:
- [ ] New Feature
- [x] Bugfix
- [ ] Documentation

## Who can review?

Anyone in the community is free to review the PR once the checks have passed. 
[Contributor guidelines](https://github.com/NVIDIA/NeMo/blob/main/CONTRIBUTING.md) contains specific people who can review PRs to various areas.

# Additional Information
* Related to # (issue)

### CODE REVIEW

**Review by MaximumEntropy — CHANGES_REQUESTED** [review]
Thanks for the fixes! :) A few comments

File: nemo/collections/nlp/modules/common/megatron/transformer.py [hunk 1]
```diff
@@ -498,7 +500,7 @@ def compute_bias(self, query_length, key_length):
         relative_position = memory_position - context_position  # shape (query_length, key_length)
         relative_position_bucket = self._relative_position_bucket(
             relative_position,  # shape (query_length, key_length)
-            bidirectional=(self.layer_type != LayerType.decoder),  # (not self.is_decoder),
+            bidirectional=(self.layer_type == LayerType.decoder),  # (self.is_decoder),
```
[the remark this answers is not in the document]
Comment (khcs) [reply nemo/collections/nlp/modules/common/megatron/transformer.py:503]: Thanks, you're right! Changed to `bidirectional=(self.attention_type != AttnMaskType.causal)`.

File: nemo/collections/nlp/modules/common/megatron/transformer.py [hunk 2]
```diff
@@ -287,6 +287,7 @@ def __init__(
         megatron_legacy=False,
         bias=True,
         headscale=False,
+        has_relative_attention_bias=False,
```
[the remark this answers is not in the document]
Comment (khcs) [reply nemo/collections/nlp/modules/common/megatron/transformer.py:290]: Yes, this is a little confusing - we compute the `position_bias` only for the first layer of Encoder/Decoder, and initialize it with zeros for the Decoder first cross-attention layer.

File: nemo/collections/nlp/modules/common/megatron/transformer.py [hunk 3]
```diff
@@ -683,9 +685,12 @@ def forward(
...
             if self.position_embedding_type == 'relative':
-                position_bias = self.compute_bias(real_seq_length, key_length)
-            else:
-                pass  # HuggingFace implementation initialize position_bias to zero when not using
+                if self.has_relative_attention_bias:
+                    position_bias = self.compute_bias(real_seq_length, key_length)
+                elif attention_mask is not None:
+                    position_bias = torch.zeros_like(attention_mask)
+                else:
+                    position_bias = torch.zeros(1, key_length, key_length)
```
[the remark this answers is not in the document]
Comment (khcs) [reply nemo/collections/nlp/modules/common/megatron/transformer.py:693]: Good point, though with the way currently is initializes them on GPU, but wouldn't hurt to add `.to(device)` to make sure ;)

[the remark this answers is not in the document]
Comment (khcs) [reply nemo/collections/nlp/modules/common/megatron/transformer.py:693]: Oh, perhaps to make sure they're on the right device for when using tensor-/pipeline- parallel?

File: nemo/collections/nlp/models/language_modeling/megatron_lm_encoder_decoder_model.py [hunk 4]
```diff
@@ -707,7 +707,7 @@ def setup(self, stage=None):
         # when using pipeline model parallel the final stage need to initialize word embeddings
         if parallel_state.get_pipeline_model_parallel_world_size() > 1:
             self.enc_dec_model.sync_initial_word_embeddings()
-            if self.enc_dec_model.position_embedding_type != 'relative':
+            if self.cfg.position_embedding_type != 'relative':
```
Comment (MaximumEntropy) [inline nemo/collections/nlp/models/language_modeling/megatron_lm_encoder_decoder_model.py:710]: To make this backward compatibile to old models without `position_embedding_type`, you will need to use `self.cfg.get('position_embedding_type', 'learned_absolute')`.

Comment (MaximumEntropy) [reply nemo/collections/nlp/models/language_modeling/megatron_lm_encoder_decoder_model.py:710]: This is why CI is failing on the BART test right now.

**Review by MaximumEntropy — CHANGES_REQUESTED** [review]
Changes for position_embedding_type's default.
